### DISCUSSION TITLE
Extendible Enums

### DISCUSSION THREAD

**Discussion by najak3d** [dotnet/csharplang#5859] [Language Ideas] [answered]
I think there should be a notation for Extendible Enums.   Under the covers, Enums are just a number.

There are many contexts where you need to use a "Key" name for something - and enums work great, because they are human-readable like a string value, but very efficient and firmly named (prevents typos).   Intellisense shows you the Named value options, and so everything works out great.

Enums even allow you to set it to values not defined by it.

So why shouldn't enums be Extendible -- or allow you to create a derivation of the Enum, that extends it, such as:

```c#
public enum BaseEnum { _None, Red, Yellow, Blue }

// Then in my Assembly:
public enum BaseEnumExt : BaseEnum { Green, Orange, Purple }
```

Perhaps the BaseEnum would require an "ExtendibleAttribute" attached to it to enable this behavior.

Without it, what is a good solution for when you want to define an enumerable type that has extendible named values?   I've kludged this by using classes that have Public Static readonly Fields that behave similarly (looks very much the same in code, but requires supplemental code for it to behave more like an enum.

Otherwise, we're stuck using String-Names for something that instead should be an Extendible Enum.

Example:  Our Map App has 20 Add-On Layers defined for it.  But the user can later add Custom Layers.   We'd like to have each later name not just be a naked string, nor do we want it to be a class.   Once the layers are defined, each Layer is defined by it's Index.  But we want that Index to also "know it's name", and function the SAME AS AN ENUM.

Currently, to try and overcome this Enum shortcoming, we've created a MapLayerName struct that looks like this:

```c#
public partial struct MapLayerName : IComparable, IComparable<MapLayerName>
{
  static public MapLayerName _UNKNOWN;
  static public MapLayerName Ownship;
  static public MapLayerName Airports;
  static public MapLayerName AirportDrawing;
  //.... 15 more

 // then we initialize these with this type code
  static private void ___LoadDefaults()
  {
      _UNKNOWN = __Create(index2Table, "UNKNOWN");
      Ownship = __Create(index2Table, "Ownship");
      Airports = __Create(index2Table, "Airports");
      AirportDrawing = __Create(index2Table, "AirportDrawing", "Airport");
      // ... and so on
  }
}
```

What we really just wanted was this:

```c#
[ExtendibleEnum]
public enum MapLayerName { UNKNOWN, Ownship, Airports, AirportDrawing, .... }

// user Code
public enum OurLayers : MapLayerName { Heliports, Targets, Grid }
```

We could have done all this with naked-strings, but then you lose Intellisense, and refactoring names/etc becomes problematic, as well as detecting/handling typos.

**Comment by svick** [comment, marked as answer]
This is a duplicate of https://github.com/dotnet/csharplang/issues/3491.

**Reply by najak3d** [reply]
I can't believe more people haven't commented on that other thread.

Extendible Enum concept is very useful, and long-overdue.

**Reply by CyrusNajmabadi** [reply]
@najak3d That might be a good way to gauge how your personal perceptions about the importance of a feature reflect the views of hte community at large :)

**Reply by najak3d** [reply]
What do you use when you need to hard-code a string-based KeyValue in C#?   Most coders I encounter simply use Strings -- which is awful, IMO.   But in their defense, C# doesn't offer an easy alternative solution, making use of string-keys, a common practice.

Some change these over to Enums - but the weakness of that, is that Enums are not extendible outside of the Assembly where they are defined.   And once they get bitten by that weakness, their response is to simply switch everything back to String-based key names.  Yuck.

What is your solution?  Have you encountered this need before?

**Reply by CyrusNajmabadi** [reply]
> Most coders I encounter 

Right. My point is taht this is useful to not draw inferences about the community at large from anecdotal sample sizes :)

**Reply by CyrusNajmabadi** [reply]
> What do you use when you need to hard-code a string-based KeyValue in C#?

I'm not sure what you mean by this.

**Reply by CyrusNajmabadi** [reply]
> Some change these over to Enums - but the weakness of that, is that Enums are not extendible outside of the Assembly where they are defined. 

I'm not sure why you need to extend in this case.  But, if you want that, you can just define constants (or static-readonly-s) for the values you want to define, then let the user define wahtever else they want.  

> Have you encountered this need before?

I'm really not sure what scenario you're talking about.  However, having APIs that predefine a set of values, but still accept any other value seems pretty normal and not strange to support without the need of having a particular enum that you 'grow'.
